sdk/eventhub/azure-eventhubs/azure/eventhub/aio/_producer_client_async.py

```diff
@@ -0,0 +1,179 @@
+
+    Example:
+        .. literalinclude:: ../samples/test_examples_eventhub_async.py
+            :start-after: [START create_eventhub_producer_client_async]
+            :end-before: [END create_eventhub_producer_client_async]
+            :language: python
+            :dedent: 4
+            :caption: Create a new instance of the EventHubProducerClient.
+    """
+
+    def __init__(self, host, event_hub_path, credential, **kwargs):
+        # type:(str, str, Union[EventHubSharedKeyCredential, EventHubSASTokenCredential, TokenCredential], Any) -> None
+        """
+        :param host: The hostname of the Event Hub.
+        :type host: str
+        :param event_hub_path: The path of the specific Event Hub to connect the client to.
+        :type event_hub_path: str
+        :param network_tracing: Whether to output network trace logs to the logger. Default
+         is `False`.
+        :type network_tracing: bool
+        :param credential: The credential object used for authentication which implements particular interface
+         of getting tokens. It accepts ~azure.eventhub.EventHubSharedKeyCredential,
+         ~azure.eventhub.EventHubSASTokenCredential, credential objects generated by the azure-identity library and
+         objects that implement get_token(self, *scopes) method.
+        :param http_proxy: HTTP proxy settings. This must be a dictionary with the following
+         keys: 'proxy_hostname' (str value) and 'proxy_port' (int value).
+         Additionally the following keys may also be present: 'username', 'password'.
+        :type http_proxy: dict[str, Any]
+        :param auth_timeout: The time in seconds to wait for a token to be authorized by the service.
+         The default value is 60 seconds. If set to 0, no timeout will be enforced from the client.
+        :type auth_timeout: float
+        :param user_agent: The user agent that needs to be appended to the built in user agent string.
+        :type user_agent: str
+        :param retry_total: The total number of attempts to redo the failed operation when an error happened. Default
+         value is 3.
+        :type retry_total: int
+        :param transport_type: The type of transport protocol that will be used for communicating with
+         the Event Hubs service. Default is ~azure.eventhub.TransportType.Amqp.
+        :type transport_type: ~azure.eventhub.TransportType
+        """
+        super(EventHubProducerClient, self).__init__(
+            host=host, event_hub_path=event_hub_path, credential=credential, **kwargs)
+        self._producers = []  # type: List[EventHubProducer]
+        self._client_lock = asyncio.Lock()  # sync the creation of self._producers
+        self._producers_locks = []  # type: List[asyncio.Lock]
+        self._max_message_size_on_link = 0
+
+    async def _init_locks_for_producers(self):
+        if not self._producers:
+            async with self._client_lock:
+                if not self._producers:
```

Comment:
We don't need to check this again

Reply:
For async tasks yes I agree in theory. If this is multi-threading, we do need to lock and test.
But let me keep this for now and do more test later to verify. It has very minimal impact. For every EventHubProducerClient instance, the lock and the second check are executed only once.